testutil/server.go

```diff
@@ -100,14 +101,57 @@ func defaultServerConfig() *TestServerConfig {
 	}
 }
 
+// Ports allocation state.
+//
+// We let the kernel choose random port number, so there's a good
+// chance the port is unused. But kernel could allocate the same
+// port twice, if agent did not have time to start listening on it.
+//
+// There is no obvious way to defend from the other processes in
+// the system, which will use OS-assigned ports, but at least it
+// will reduce duplicated ports shared by the agents.
+var nmap = make(map[int]struct{})
+var nmapmu sync.Mutex
+
 // randomPort asks the kernel for a random port to use.
 func randomPort() int {
-	l, err := net.Listen("tcp", "127.0.0.1:0")
-	if err != nil {
-		panic(err)
+	makeport := func() (int, bool) {
```

Comment:
Right, I understand this play with ports allocation is necessary to speed up the tests invocation only, definitely it is not targeted for production usage.

But I did not precisely get the point with a global variables that are not shared across multiple packages (I assume you are talking about the **testing** packages).

If so, then ```go test``` command will *sequentially* (one package by one) compile a standalone binary for each testing package and launch the tests in parallel only within the *single* package:
```text
go test --help
...
        -parallel n
            Allow parallel execution of test functions that call t.Parallel.
            The value of this flag is the maximum number of tests to run
            simultaneously; by default, it is set to the value of GOMAXPROCS.
            Note that -parallel only applies within a single test binary.
            The 'go test' command may run tests for different packages
            in parallel as well, according to the setting of the -p flag
            (see 'go help build').
...
```

So this map will be shared within a single binary. I also agree that proposed changes does not solve the problem (because there is clearly no solution). These changes to my mind reduce the probability of the case, when we allocate two same ports while testing the single package.